docs/design/ftlib.md

```diff
@@ -19,32 +19,43 @@ Additionally, since most communication libraries require rank assigned to each w
+For each batch of data, the model in each worker will be fed with training data, producing loss by forwarding and gradients by backward propagation. With local gradients ready on each worker, the allreduce API in FTLib will be called to average the gradients across all workers. As the operation is blocking, it indicates all parameters on this process is ready to be updated if allreduce API returns success.
 
-Before launching any communication operations, the FTLib will lock the process and unlock after finished or aborted. This design prevents the initialized flag from being modified by `Consensus.new_member_join` on other thread during the average phase.
+Otherwise, the training process will abort or continue according to different returned status.
 
-The initialized flag of a FTLib will be reset (to `False`) under three occasions:
+The entire step described above can be wrapped into a high-level customized API with FTLib, which hides the latency of communication by overlapping computation and communication.
 
-1. New workers report joining in
-2. Collective-communication operations fail or time out
-3. *Liveness probe fails for existing workers*
 
-Here we assume the third case is or will eventually lead to the second one. Thusly, only the first and second occasions are handled explicitly in FTLib.
 
-*In the next design, we will consider hiding communication APIs of FTLib by wrapping model and optimizer, which hooks the communication operations for each parameter to its gradient calculation function. In this way, the optimizer will only need to check if all parameters hook are executed before updating weights.*
+## Passive or Active Mode
+
+With different consensus, FTLib wraps communication APIs with two modes:
+
+- Active Mode: check member list consensus in `initialized()` and `skip_allreduce()` APIs  
+- Passive Mode: not perform the check described in activte mode
+
+Before launching any communication operations, the FTLib will lock the process and unlock after finished or aborted. This design prevents the initialized flag from being modified by `Consensus.new_member_join`.
 
-**The next design will have the communication and backward computation overlapped, reducing the overall training time.**
+For details about passive and active mode in consensus, please refer the [consensus design doc](./consensus.md).
 
-## FTLib Flow
+## FTLib Flow with Low-Level APIs
 
-The flow chart below describes how FTLib works internally.
+The flow chart below describes how FTLib works internally when low-level APIs are used in distributed training.
 
 ![ftlib_flow](../imgs/ftlib_flow.png)
 
 *Fig 3. The Internal Mechanism of FTLib*
 
+The initialized flag of a FTLib will be reset (to `False`) under three occasions:
+
+1. New workers report joining in
+2. Collective-communication operations fail or time out
+3. *Liveness probe fails for existing workers*
```

Comment:
May need explain how the liveness probe is being done for the workers

Reply:
Sure, I added the following paragraph.

There are different approaches for liveness probe. *(For example, we force a `clean_ip` function for each worker at exit and scan the shared storage to check the existence of ip files as liveness probe. However, in the gossip-based consensus protocol, each worker will pull/push gossip messages of liveness from/to random nodes as liveness probe.)*